File: app/src/main/java/org/tvheadend/tvhclient/ui/features/settings/SettingsListConnectionsFragment.java
package org.tvheadend.tvhclient.ui.features.settings;


import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.ActionMode;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.ListView;

import com.afollestad.materialdialogs.MaterialDialog;

import org.tvheadend.tvhclient.MainApplication;
import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.data.repository.AppRepository;
import org.tvheadend.tvhclient.data.service.HtspService;
import org.tvheadend.tvhclient.domain.entity.Connection;
import org.tvheadend.tvhclient.ui.base.callbacks.BackPressedInterface;
import org.tvheadend.tvhclient.ui.base.callbacks.ToolbarInterface;
import org.tvheadend.tvhclient.ui.features.startup.SplashActivity;
import org.tvheadend.tvhclient.util.tasks.WakeOnLanTask;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.ListFragment;
import androidx.lifecycle.ViewModelProviders;
import timber.log.Timber;

public class SettingsListConnectionsFragment extends ListFragment implements BackPressedInterface, ActionMode.Callback {

    private ToolbarInterface toolbarInterface;
    private ConnectionListAdapter connectionListAdapter;
    private ActionMode actionMode;
    private AppCompatActivity activity;
    @Inject
    protected AppRepository appRepository;
    private ConnectionViewModel viewModel;

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        MainApplication.getComponent().inject(this);

        activity = (AppCompatActivity) getActivity();
        if (activity instanceof ToolbarInterface) {
            toolbarInterface = (ToolbarInterface) activity;
        }
        toolbarInterface.setTitle(getString(R.string.settings));

        connectionListAdapter = new ConnectionListAdapter(activity);
        setListAdapter(connectionListAdapter);
        getListView().setChoiceMode(ListView.CHOICE_MODE_SINGLE);
        setHasOptionsMenu(true);

        viewModel = ViewModelProviders.of(activity).get(ConnectionViewModel.class);
        viewModel.setConnectionHasChanged(false);
        viewModel.getAllConnections().observe(activity, connections -> {
            if (connections != null) {
                connectionListAdapter.clear();
                connectionListAdapter.addAll(connections);
                connectionListAdapter.notifyDataSetChanged();
                toolbarInterface.setSubtitle(activity.getResources().getQuantityString(
                        R.plurals.number_of_connections,
                        connectionListAdapter.getCount(),
                        connectionListAdapter.getCount()));
            }
        });
    }

    @Override
    public void onListItemClick(@NonNull ListView l, @NonNull View v, int position, long id) {
        if (actionMode == null) {
            getListView().setItemChecked(position, true);
            getListView().setSelected(true);
            startActionMode();
        }
    }

    private void startActionMode() {
        actionMode = activity.startActionMode(this);
        if (actionMode != null) {
            actionMode.invalidate();
        }
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.connection_add_options_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_add:
                Intent intent = new Intent(activity, SettingsActivity.class);
                intent.putExtra("setting_type", "add_connection");
                startActivity(intent);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
        int position = getListView().getCheckedItemPosition();
        Connection connection = connectionListAdapter.getItem(position);
        if (connection == null) {
            return false;
        }
        Intent intent;
        switch (item.getItemId()) {
            case R.id.menu_set_active:
                connection.setActive(true);
                appRepository.getConnectionData().updateItem(connection);
                viewModel.setConnectionHasChanged(true);
                mode.finish();
                return true;

            case R.id.menu_set_not_active:
                connection.setActive(false);
                appRepository.getConnectionData().updateItem(connection);
                mode.finish();
                return true;

            case R.id.menu_edit:
                intent = new Intent(activity, SettingsActivity.class);
                intent.putExtra("setting_type", "edit_connection");
                intent.putExtra("connection_id", connection.getId());
                startActivity(intent);
                mode.finish();
                return true;

            case R.id.menu_send_wol:
                new WakeOnLanTask(activity, connection).execute();
                mode.finish();
                return true;

            case R.id.menu_delete:
                new MaterialDialog.Builder(activity)
                        .content(getString(R.string.delete_connection, connection.getName()))
                        .positiveText(getString(R.string.delete))
                        .negativeText(getString(R.string.cancel))
                        .onPositive((dialog, which) -> appRepository.getConnectionData().removeItem(connection))
                        .onNegative((dialog, which) -> dialog.cancel())
                        .show();
                mode.finish();
                return true;

            default:
                return false;
        }
    }

    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
        // Inflate a menu resource providing context menu items
        MenuInflater inflater = mode.getMenuInflater();
        inflater.inflate(R.menu.connection_list_options_menu, menu);
        return true;
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {
        actionMode = null;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
        // Get the currently selected program from the list
        int position = getListView().getCheckedItemPosition();
        final Connection connection = connectionListAdapter.getItem(position);
        if (connection != null) {
            // Show or hide the wake on LAN menu item
            menu.getItem(0).setVisible(!TextUtils.isEmpty(connection.getWolMacAddress()));
            // Show or hide the activate / deactivate menu items
            menu.getItem(1).setVisible(!connection.isActive());
            menu.getItem(2).setVisible(connection.isActive());

            mode.setTitle(connection.getName());
        }
        return true;
    }

    @Override
    public void onBackPressed() {
        if (viewModel.getActiveConnectionId() < 0) {
            new MaterialDialog.Builder(activity)
                    .title(R.string.dialog_title_disconnect_from_server)
                    .content(R.string.dialog_content_disconnect_from_server)
                    .positiveText(R.string.disconnect)
                    .onPositive((dialog, which) -> reconnect())
                    .show();
        } else if (viewModel.getConnectionHasChanged()) {
            new MaterialDialog.Builder(activity)
                    .title(R.string.dialog_title_connection_changed)
                    .content(R.string.dialog_content_connection_changed)
                    .positiveText(R.string.connect)
                    .onPositive((dialog, which) -> reconnect())
                    .show();
        } else {
            viewModel.setConnectionHasChanged(false);
            activity.finish();
        }
    }

    /**
     * Save the information that a new sync is required and stop
     * the service so it can refresh the active connection.
     * Then restart the application to show the startup fragment
     */
    private void reconnect() {
        Timber.d("Reconnecting to server, new initial sync will be done");
        activity.stopService(new Intent(activity, HtspService.class));

        if (viewModel.getActiveConnectionId() >= 0) {
            Connection connection = appRepository.getConnectionData().getItemById(viewModel.getActiveConnectionId());
            if (connection != null) {
                connection.setSyncRequired(true);
                connection.setLastUpdate(0);
                appRepository.getConnectionData().updateItem(connection);
            }
        }

        Intent intent = new Intent(activity, SplashActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        activity.startActivity(intent);

        viewModel.setConnectionHasChanged(false);
        activity.finish();
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/settings/SettingsListConnectionsFragment.kt
package org.tvheadend.tvhclient.ui.features.settings


import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.*
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.afollestad.materialdialogs.MaterialDialog
import org.tvheadend.data.entity.Connection
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.databinding.RecyclerviewFragmentBinding
import org.tvheadend.tvhclient.service.ConnectionService
import org.tvheadend.tvhclient.ui.common.WakeOnLanTask
import org.tvheadend.tvhclient.ui.common.interfaces.BackPressedInterface
import org.tvheadend.tvhclient.ui.common.interfaces.RecyclerViewClickInterface
import org.tvheadend.tvhclient.ui.common.interfaces.ToolbarInterface
import org.tvheadend.tvhclient.ui.features.MainActivity

class SettingsListConnectionsFragment : Fragment(), BackPressedInterface, ActionMode.Callback, RecyclerViewClickInterface {

    private lateinit var binding: RecyclerviewFragmentBinding
    private var activeConnectionId: Int = -1
    private var connectionHasChanged: Boolean = false
    private lateinit var toolbarInterface: ToolbarInterface
    private lateinit var recyclerViewAdapter: ConnectionRecyclerViewAdapter
    private lateinit var settingsViewModel: SettingsViewModel

    private var actionMode: ActionMode? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = RecyclerviewFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        settingsViewModel = ViewModelProvider(activity as SettingsActivity)[SettingsViewModel::class.java]

        if (activity is ToolbarInterface) {
            toolbarInterface = activity as ToolbarInterface
            toolbarInterface.setTitle(getString(R.string.pref_connections))
        }

        recyclerViewAdapter = ConnectionRecyclerViewAdapter(this, viewLifecycleOwner)
        binding.recyclerView.layoutManager = LinearLayoutManager(activity)
        binding.recyclerView.adapter = recyclerViewAdapter

        settingsViewModel.connectionListLiveData.observe(viewLifecycleOwner) { connections ->
            if (connections != null) {
                recyclerViewAdapter.addItems(connections)
                context?.let {
                    toolbarInterface.setSubtitle(
                        it.resources.getQuantityString(
                            R.plurals.number_of_connections,
                            recyclerViewAdapter.itemCount,
                            recyclerViewAdapter.itemCount
                        )
                    )
                }
            }
        }
        settingsViewModel.activeConnectionLiveData.observe(viewLifecycleOwner) { connection ->
            connectionHasChanged = connection != null && connection.id != settingsViewModel.connectionToEdit.id
            activeConnectionId = connection?.id ?: -1
        }

        setHasOptionsMenu(true)
    }

    private fun startActionMode() {
        actionMode = activity?.startActionMode(this)
        actionMode?.invalidate()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.connection_add_options_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_add_connection -> addConnection()
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        val ctx = context ?: return false
        val position = recyclerViewAdapter.selectedPosition
        if (recyclerViewAdapter.itemCount <= position) return false

        val connection = recyclerViewAdapter.getItem(position) ?: return false
        return when (item.itemId) {
            R.id.menu_set_connection_active -> setConnectionActiveOrInactive(connection, mode, true)
            R.id.menu_set_connection_not_active -> setConnectionActiveOrInactive(connection, mode, false)
            R.id.menu_edit_connection -> editConnection(connection, mode)
            R.id.menu_send_wol -> {
                WakeOnLanTask(lifecycleScope, ctx, connection)
                mode.finish()
                true
            }
            R.id.menu_delete_connection -> deleteConnection(ctx, connection, mode)
            else -> false
        }
    }

    private fun addConnection(): Boolean {
        settingsViewModel.setNavigationMenuId("add_connection")
        return true
    }

    private fun editConnection(connection: Connection, mode: ActionMode): Boolean {
        settingsViewModel.connectionIdToBeEdited = connection.id
        settingsViewModel.setNavigationMenuId("edit_connection")
        mode.finish()
        return true
    }

    private fun setConnectionActiveOrInactive(connection: Connection, mode: ActionMode, active: Boolean): Boolean {
        connection.isActive = active
        settingsViewModel.updateConnection(connection)
        mode.finish()
        return true
    }

    private fun deleteConnection(context: Context, connection: Connection, mode: ActionMode): Boolean {
        MaterialDialog(context).show {
            message(text = getString(R.string.delete_connection, connection.name))
            positiveButton(R.string.delete) {
                settingsViewModel.removeConnection(connection)
            }
            negativeButton(R.string.cancel) {
                cancel()
            }
        }
        mode.finish()
        return true
    }

    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        mode.menuInflater.inflate(R.menu.connection_list_options_menu, menu)
        return true
    }

    override fun onDestroyActionMode(mode: ActionMode) {
        actionMode = null
    }

    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
        // Get the currently selected program from the list
        val position = recyclerViewAdapter.selectedPosition
        val connection = recyclerViewAdapter.getItem(position)
        if (connection != null) {
            // Show or hide the wake on LAN menu item
            menu.getItem(0).isVisible = connection.isWolEnabled && !connection.wolMacAddress.isNullOrEmpty()
            // Show or hide the activate / deactivate menu items
            menu.getItem(1).isVisible = !connection.isActive
            menu.getItem(2).isVisible = connection.isActive

            mode.title = connection.name
        }
        return true
    }

    override fun onBackPressed() {
        when {
            activeConnectionId < 0 -> context?.let {
                MaterialDialog(it).show {
                    title(R.string.disconnect_from_server)
                    message(R.string.no_active_connection)
                    positiveButton(R.string.disconnect) {
                        updateConnectionAndRestartApplication()
                    }
                }
            }
            connectionHasChanged -> context?.let {
                MaterialDialog(it).show {
                    title(R.string.connect_to_new_server)
                    message(R.string.connection_changed)
                    positiveButton(R.string.connect) {
                        updateConnectionAndRestartApplication()
                    }
                }
            }
            else -> {
                (activity as RemoveFragmentFromBackstackInterface).removeFragmentFromBackstack()
            }
        }
    }

    private fun updateConnectionAndRestartApplication() {
        context?.stopService(Intent(context, ConnectionService::class.java))
        val intent = Intent(context, MainActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        context?.startActivity(intent)
    }

    override fun onClick(view: View, position: Int) {
        actionMode?.finish()
        if (actionMode == null) {
            recyclerViewAdapter.setPosition(position)
            startActionMode()
        }
    }

    override fun onLongClick(view: View, position: Int): Boolean {
        return true
    }
}
